File: packages/react-native-webcapsule/android/src/main/java/com/webcapsule/reactnative/runtime/UpdateIndex.kt
package com.webcapsule.reactnative.runtime

import java.net.URI
import java.nio.charset.StandardCharsets
import java.util.Base64

data class UpdateRelease(
  val version: String,
  val url: URI,
  val sha256: String,
  val size: Long,
  val minimumRuntimeVersion: String,
)

data class VerifiedUpdateIndex(
  val capsuleId: String,
  val channel: String,
  val keyId: String,
  val releases: List<UpdateRelease>,
)

object UpdateIndexVerifier {
  private val capsuleIdPattern = Regex("^[a-z0-9]+(?:[.-][a-z0-9]+)+$")
  private val channelPattern = Regex("^[a-z0-9][a-z0-9._-]{0,63}$")
  private val keyIdPattern = Regex("^[A-Za-z0-9][A-Za-z0-9._-]{0,127}$")
  private val hashPattern = Regex("^[0-9a-f]{64}$")
  private val signaturePattern = Regex("^[A-Za-z0-9+/]{86}==$")
  private val rootFields = setOf("schemaVersion", "capsuleId", "channel", "releases", "keyId", "signature")
  private val releaseFields = setOf("version", "url", "sha256", "size", "minimumRuntimeVersion")

  fun verify(
    bytes: ByteArray,
    expectedCapsuleId: String,
    expectedChannel: String,
    publicKeys: Map<String, String>,
  ): VerifiedUpdateIndex {
    if (bytes.isEmpty() || bytes.size > 1024 * 1024) fail(WebCapsuleErrorCode.LIMIT_EXCEEDED, "Update index size is invalid")
    val root = (StrictJson.parse(bytes) as? Map<*, *>) ?: invalid("Update index must be an object")
    exact(root, rootFields, "Update index")
    if (integer(root, "schemaVersion") != 1L) invalid("Unsupported update index schema")
    val capsuleId = string(root, "capsuleId")
    val channel = string(root, "channel")
    val keyId = string(root, "keyId")
    if (!capsuleIdPattern.matches(capsuleId)) fail(WebCapsuleErrorCode.INVALID_CAPSULE_ID, "Invalid capsule ID")
    if (!channelPattern.matches(channel)) invalid("Invalid channel")
    if (!keyIdPattern.matches(keyId)) fail(WebCapsuleErrorCode.INVALID_KEY_ID, "Invalid key ID")
    if (capsuleId != expectedCapsuleId || channel != expectedChannel) {
      fail(WebCapsuleErrorCode.INVALID_UPDATE_INDEX, "Update index identity differs")
    }
    val releases = (root["releases"] as? List<*>)?.map { parseRelease(it) } ?: invalid("releases must be an array")
    if (releases.isEmpty()) invalid("At least one release is required")
    if (releases.zipWithNext().any { ManifestParser.compareVersions(it.first.version, it.second.version) <= 0 }) {
      fail(WebCapsuleErrorCode.INVALID_ORDER, "Releases must have unique descending SemVer precedence")
    }
    val encoded = string(root, "signature")
    if (!signaturePattern.matches(encoded)) fail(WebCapsuleErrorCode.INVALID_SIGNATURE, "Invalid update index signature encoding")
    val signature = try { Base64.getDecoder().decode(encoded) } catch (error: Exception) {
      fail(WebCapsuleErrorCode.INVALID_SIGNATURE, "Invalid update index signature", error)
    }
    if (signature.size != 64) fail(WebCapsuleErrorCode.INVALID_SIGNATURE, "Invalid update index signature length")
    val unsigned = LinkedHashMap<Any?, Any?>(root).also { it.remove("signature") }
    val canonical = try {
      StrictJson.canonicalize(json(unsigned).toByteArray(StandardCharsets.UTF_8))
    } catch (error: Exception) {
      invalid("Cannot canonicalize update index", error)
    }
    val pem = publicKeys[keyId] ?: fail(WebCapsuleErrorCode.KEY_ID_MISMATCH, "No exact trusted update key ID")
    SignatureVerifier.verifyUpdateIndex(canonical, signature, pem)
    return VerifiedUpdateIndex(capsuleId, channel, keyId, releases)
  }

  fun select(index: VerifiedUpdateIndex, runtimeVersion: String, highestSeenVersion: String, blockedVersions: Set<String>): UpdateRelease? {
    ManifestParser.assertVersion(runtimeVersion)
    ManifestParser.assertVersion(highestSeenVersion)
    return index.releases.firstOrNull {
      ManifestParser.compareVersions(it.minimumRuntimeVersion, runtimeVersion) <= 0 &&
        ManifestParser.compareVersions(it.version, highestSeenVersion) > 0 && it.version !in blockedVersions
    }
  }

  private fun parseRelease(value: Any?): UpdateRelease {
    val map = value as? Map<*, *> ?: invalid("Release must be an object")
    exact(map, releaseFields, "Release")
    val version = string(map, "version"); ManifestParser.assertVersion(version)
    val minimum = string(map, "minimumRuntimeVersion"); ManifestParser.assertVersion(minimum)
    val digest = string(map, "sha256")
    if (!hashPattern.matches(digest)) fail(WebCapsuleErrorCode.INVALID_HASH, "Invalid release SHA-256")
    val size = integer(map, "size")
    if (size > 100L * 1024 * 1024) fail(WebCapsuleErrorCode.LIMIT_EXCEEDED, "Release size exceeds capsule limit")
    val uri = strictHttps(string(map, "url"))
    return UpdateRelease(version, uri, digest, size, minimum)
  }

  fun strictHttps(value: String): URI {
    val uri = try { URI(value) } catch (error: Exception) { fail(WebCapsuleErrorCode.INVALID_URL, "Invalid HTTPS URL", error) }
    if (!uri.isAbsolute || uri.scheme != "https" || uri.host == null || uri.rawAuthority == null ||
      uri.userInfo != null || uri.rawFragment != null || uri.port != -1
    ) {
      fail(WebCapsuleErrorCode.INVALID_URL, "URL must be absolute HTTPS without userinfo, fragment, or explicit port")
    }
    return uri
  }

  private fun exact(map: Map<*, *>, fields: Set<String>, label: String) { if (map.keys.toSet() != fields) invalid("$label fields differ") }
  private fun string(map: Map<*, *>, key: String) = map[key] as? String ?: invalid("$key must be a string")
  private fun integer(map: Map<*, *>, key: String) = (map[key] as? Long)?.takeIf { it >= 0 } ?: invalid("$key must be a non-negative integer")
  private fun invalid(message: String, cause: Throwable? = null): Nothing = fail(WebCapsuleErrorCode.INVALID_UPDATE_INDEX, message, cause)

  private fun json(value: Any?): String = when (value) {
    null -> "null"
    is String -> quote(value)
    is Long, is Int -> value.toString()
    is Boolean -> value.toString()
    is List<*> -> value.joinToString(",", "[", "]") { json(it) }
    is Map<*, *> -> value.entries.joinToString(",", "{", "}") { json(it.key as String) + ":" + json(it.value) }
    else -> invalid("Unsupported update index value")
  }

  private fun quote(value: String) = buildString {
    append('"')
    value.forEach { character ->
      when (character) {
        '"' -> append("\\\"")
        '\\' -> append("\\\\")
        '\b' -> append("\\b")
        '\u000c' -> append("\\f")
        '\n' -> append("\\n")
        '\r' -> append("\\r")
        '\t' -> append("\\t")
        else -> if (character.code < 0x20) append("\\u%04x".format(character.code)) else append(character)
      }
    }
    append('"')
  }
}

File: packages/react-native-webcapsule/ios/Sources/WebCapsuleCore/UpdateIndex.swift
import Foundation

struct UpdateRelease: Equatable, Sendable {
    let version: String
    let url: URL
    let sha256: String
    let size: Int64
    let minimumRuntimeVersion: String
}

struct VerifiedUpdateIndex: Equatable, Sendable {
    let capsuleId: String
    let channel: String
    let keyId: String
    let releases: [UpdateRelease]
}

enum UpdateIndexVerifier {
    private static let capsuleLimit: Int64 = 100 * 1024 * 1024
    private static let indexLimit = 1024 * 1024
    private static let rootFields: Set<String> = [
        "schemaVersion", "capsuleId", "channel", "releases", "keyId", "signature",
    ]
    private static let releaseFields: Set<String> = [
        "version", "url", "sha256", "size", "minimumRuntimeVersion",
    ]

    static func verify(
        _ data: Data,
        expectedCapsuleId: String,
        expectedChannel: String,
        publicKeys: [String: String]
    ) throws -> VerifiedUpdateIndex {
        guard !data.isEmpty, data.count <= indexLimit else {
            throw WebCapsuleError(code: .limitExceeded, message: "Update index size is invalid")
        }
        let parsed = try StrictJSON.parse(data)
        guard case let .object(root) = parsed else { throw invalid("Update index must be an object") }
        try exact(root, fields: rootFields, label: "Update index")
        guard case let .integer(schemaVersion) = try required(root, "schemaVersion"), schemaVersion == 1,
              case let .string(capsuleId) = try required(root, "capsuleId"),
              case let .string(channel) = try required(root, "channel"),
              case let .array(releaseValues) = try required(root, "releases"),
              case let .string(keyId) = try required(root, "keyId"),
              case let .string(encodedSignature) = try required(root, "signature") else {
            throw invalid("Update index field type or schema is invalid")
        }
        do { try CapsuleManifestParser.validateCapsuleID(capsuleId) }
        catch { throw WebCapsuleError(code: .invalidCapsuleID, message: "Invalid capsule ID") }
        guard validChannel(channel) else { throw invalid("Invalid channel") }
        do { try CapsuleManifestParser.validateKeyID(keyId) }
        catch { throw WebCapsuleError(code: .invalidKeyID, message: "Invalid key ID") }
        guard capsuleId == expectedCapsuleId, channel == expectedChannel else {
            throw invalid("Update index identity differs")
        }
        guard !releaseValues.isEmpty else { throw invalid("At least one release is required") }
        let releases = try releaseValues.map(parseRelease)
        for pair in zip(releases, releases.dropFirst()) {
            guard try SemanticVersion.compare(pair.0.version, pair.1.version) == .orderedDescending else {
                throw WebCapsuleError(
                    code: .invalidOrder,
                    message: "Releases must have unique descending SemVer precedence"
                )
            }
        }
        let signature = try parseSignature(encodedSignature)
        let unsigned = StrictJSONValue.object(StrictJSONObject(
            entries: root.entries.filter { $0.key != "signature" }
        ))
        try SignedManifestVerifier.verifyUpdateIndex(
            canonicalIndex: CanonicalJSON.serialize(unsigned),
            signature: signature,
            keyId: keyId,
            publicKeys: publicKeys
        )
        return VerifiedUpdateIndex(
            capsuleId: capsuleId,
            channel: channel,
            keyId: keyId,
            releases: releases
        )
    }

    static func select(
        _ index: VerifiedUpdateIndex,
        runtimeVersion: String,
        highestSeenVersion: String,
        blockedVersions: Set<String>
    ) throws -> UpdateRelease? {
        try SemanticVersion.validate(runtimeVersion)
        try SemanticVersion.validate(highestSeenVersion)
        for release in index.releases where
            try SemanticVersion.compare(release.minimumRuntimeVersion, runtimeVersion) != .orderedDescending
                && SemanticVersion.compare(release.version, highestSeenVersion) == .orderedDescending
                && !blockedVersions.contains(release.version) {
            return release
        }
        return nil
    }

    static func strictHTTPS(_ value: String) throws -> URL {
        guard !value.isEmpty,
              let components = URLComponents(string: value),
              components.scheme == "https",
              components.host != nil,
              !components.host!.isEmpty,
              components.user == nil,
              components.password == nil,
              components.fragment == nil,
              components.port == nil,
              !hasExplicitPort(value),
              let url = components.url,
              url.baseURL == nil else {
            throw WebCapsuleError(
                code: .invalidURL,
                message: "URL must be absolute HTTPS without userinfo, fragment, or explicit port"
            )
        }
        return url
    }

    private static func hasExplicitPort(_ value: String) -> Bool {
        guard let authorityStart = value.range(of: "://")?.upperBound else { return false }
        let remainder = value[authorityStart...]
        let authority = remainder.prefix { $0 != "/" && $0 != "?" && $0 != "#" }
        if authority.hasPrefix("[") {
            guard let closing = authority.firstIndex(of: "]") else { return true }
            return authority.index(after: closing) != authority.endIndex
        }
        return authority.contains(":")
    }

    static func validateChannel(_ value: String) throws {
        guard validChannel(value) else { throw invalid("Invalid channel") }
    }

    private static func parseRelease(_ value: StrictJSONValue) throws -> UpdateRelease {
        guard case let .object(object) = value else { throw invalid("Release must be an object") }
        try exact(object, fields: releaseFields, label: "Release")
        guard case let .string(version) = try required(object, "version"),
              case let .string(urlString) = try required(object, "url"),
              case let .string(sha256) = try required(object, "sha256"),
              case let .integer(size) = try required(object, "size"),
              case let .string(minimumRuntimeVersion) = try required(object, "minimumRuntimeVersion") else {
            throw invalid("Release field type is invalid")
        }
        try SemanticVersion.validate(version)
        try SemanticVersion.validate(minimumRuntimeVersion)
        guard VersionRecordCodec.isLowercaseSHA256(sha256) else {
            throw WebCapsuleError(code: .invalidHash, message: "Invalid release SHA-256")
        }
        guard size >= 0 else { throw invalid("Release size must be non-negative") }
        guard size <= capsuleLimit else {
            throw WebCapsuleError(code: .limitExceeded, message: "Release size exceeds capsule limit")
        }
        return UpdateRelease(
            version: version,
            url: try strictHTTPS(urlString),
            sha256: sha256,
            size: size,
            minimumRuntimeVersion: minimumRuntimeVersion
        )
    }

    private static func parseSignature(_ value: String) throws -> Data {
        let bytes = Array(value.utf8)
        guard bytes.count == 88,
              bytes.suffix(2).elementsEqual([0x3D, 0x3D]),
              bytes.dropLast(2).allSatisfy(isBase64Character),
              let signature = Data(base64Encoded: value, options: []),
              signature.count == 64,
              signature.base64EncodedString() == value else {
            throw WebCapsuleError(code: .invalidSignature, message: "Invalid update index signature encoding")
        }
        return signature
    }

    private static func exact(_ object: StrictJSONObject, fields: Set<String>, label: String) throws {
        guard object.entries.count == fields.count, Set(object.entries.map(\.key)) == fields else {
            throw invalid("\(label) fields differ")
        }
    }

    private static func required(_ object: StrictJSONObject, _ key: String) throws -> StrictJSONValue {
        guard let value = object[key] else { throw invalid("Update index field is missing") }
        return value
    }

    private static func validChannel(_ value: String) -> Bool {
        let bytes = Array(value.utf8)
        guard (1...64).contains(bytes.count), let first = bytes.first, isLowerAlphaNumeric(first) else {
            return false
        }
        return bytes.allSatisfy { isLowerAlphaNumeric($0) || $0 == 0x2E || $0 == 0x5F || $0 == 0x2D }
    }

    private static func isLowerAlphaNumeric(_ byte: UInt8) -> Bool {
        (0x61...0x7A).contains(byte) || (0x30...0x39).contains(byte)
    }

    private static func isBase64Character(_ byte: UInt8) -> Bool {
        (0x41...0x5A).contains(byte) || (0x61...0x7A).contains(byte)
            || (0x30...0x39).contains(byte) || byte == 0x2B || byte == 0x2F
    }

    private static func invalid(_ message: String) -> WebCapsuleError {
        WebCapsuleError(code: .invalidUpdateIndex, message: message)
    }
}
